javascript/ql/src/semmle/javascript/dataflow/internal/InterProceduralTypeInference.qll

```diff
@@ -230,3 +232,47 @@ private class TypeInferredCalleeWithAnalyzedReturnFlow extends CallWithNonLocalA
 
   override AnalyzedFunction getACallee() { result = fun }
 }
+
+/**
+ * Holds if `call` uses `receiver` as its only receiver value.
+ */
+pragma[noinline]
+private predicate hasDefiniteReceiver(
+  DataFlow::MethodCallNode call, CapturedSource receiver
+) {
+  call = receiver.getAMethodCall() and
+  exists (DataFlow::AnalyzedNode receiverNode, AbstractValue abstractCapturedReceiver |
+    receiverNode = call.getReceiver() and
+    not receiverNode.getALocalValue().isIndefinite(_) and
+    abstractCapturedReceiver = receiver.analyze().getALocalValue() and
+    forall(DataFlow::AbstractValue v |
+      receiverNode.getALocalValue() = v |
+      v = abstractCapturedReceiver
+    )
+  )
+}
+
+/**
+ * Enables inter-procedural type inference for the return value of a
+ * method call to a flow-insensitively type-inferred callee.
+ */
+class TypeInferredMethodWithAnalyzedReturnFlow extends CallWithNonLocalAnalyzedReturnFlow {
+  DataFlow::FunctionNode fun;
+
+  TypeInferredMethodWithAnalyzedReturnFlow() {
+    exists(CapturedSource s, DataFlow::PropWrite w, string name |
+      this.(DataFlow::MethodCallNode).getMethodName() = name and
+      s.hasOwnProperty(name) and
+      hasDefiniteReceiver(this, s) and
+      w = s.getAPropertyWrite() and
+      fun.flowsTo(w.getRhs()) and
```

Comment:
I assume we guard somewhere else against the case that the RHS is a function that we cannot track?

Reply:
Yes, a `LocalObject` object can not have an unknown method, because that method could expose the object through `this`.